### ISSUE TITLE
Getting Swahili videos to offline server

### ISSUE THREAD

**Issue by Kevsy** [learningequality/ka-lite#5294]
## Summary

Installing the Swahili content pack only changes the user interface language, but videos are still in English
## System information

_Please specify the KA Lite version you were using and your operating system!_
- Operating system: Windows Server 2012
- Version: 0.16.8
- Browser: Any
## Traceback or relevant snippet from server.log

```
INSERT TRACEBACK, LOG MESSAGES ETC. HERE
```
## How to reproduce
1. Install English videos
2. Install Swahili content pack using the instructions for a server with no Internet connection
3. Select any video

Note that https://sw.khanacademy.org seems to have all the videos, but installing the content pack does not pull these into KA Lite.
## Screenshots

(awaiting these)
## Real-life consequences (anything community should be aware, for instance how it affects your deployment)

Swahili is a mandatory requirement for primary school in Tanzania, so this is critical for launch.

Many thanks as always !

**Comment by benjaoming** [comment]
Hi @Kevsy 

Thanks for reporting this. The only thing that looks remotely correct about the Swahili content pack is its "beta: True" flag :)

I've added some info on the project that builds the content packs.

I'm very convinced that Swahili should be prioritized.

If you want to get things running quickly, you could look at KA Lite 0.15, which was before we changed the format for language data.. however, it naturally has other drawbacks.

**Comment by julianharty** [comment]
I agree that Swahili is worth prioritizing in terms of getting all the
content into ka-lite. I've done some work (including funding some of the
translation work) and am willing to help. What I can't do is work unaided
to sort out the many details that need to be addressed for Swahili to be
adequately implemented in ka-lite. I do know various organisations who are
keen to use ka-lite with Swahili as soon as it's a strong beta. What I have
yet to meet are people who are willing and able to actively participate
and/or fund the development and project work needed - which means that the
project depends a great deal on the good will and availability of Ben, etc.

Ciao for now


Julian

**Comment by benjaoming** [comment]
@mrpau-eduard you're not showing up on the list of selectable assignees

**Comment by benjaoming** [comment]
Thanks @julianharty - we are very keen on having KA Lite translated to Swahili, and most of the efforts (excluding the interface translations) done building a content pack for KA Lite will also be utilized for [Kolibri](https://learningequality.org/kolibri/).

**Comment by mrpau-eduard** [comment]
@benjaoming I get it now. I need to be a memmeber of [learningequality](https://github.com/learningequality) in order to be assigned with a task. 
I am using the company account @benjaoming .

**Comment by benjaoming** [comment]
Hi @Kevsy and @julianharty 

@mrpau-eduard and @mrpau-richard have built a Swahili content pack, and it's uploaded and live for the 0.16 series - are you able to have a look and give some feedback?

**Comment by Kevsy** [comment]
Many thanks everyone! I will pass it to the Tanzania team and keep you posted. 

Cheers
Kevin

**Comment by julianharty** [comment]
Very well done. I'll check in the next few days (I'm flying today to the
USA - and visiting the team next Monday :)

Julian

**Comment by Kevsy** [comment]
Thanks @benjaoming , @mrpau-eduard & @mrpau-richard for your hard work! However I'm not getting Kiswahili in the menu tree - rather it's blank. I don't think it's a character encoding issue because there is some Kiswahili in the UI, and the videos appear in the search results.

`$ kalite --version`
`0.16.9`

` ~/Code/kalite$ ls -l sw.zip`
`-rw-r--r--@ 1 ###### ####  2881058 28 Sep 13:06 sw.zip`

`$ kalite manage retrievecontentpack local sw sw.zip`
`[INFO] [2016-09-28 16:25:35,277] kalite: Installing a local content pack.`
`[INFO] [2016-09-28 16:25:35,286] kalite: Moving content files to the right place.`
`[WARNING] [2016-09-28 16:25:35,592] kalite: No sw_icu.js file found in locale_path` `/Users/####/.kalite/locale`
`[WARNING] [2016-09-28 16:25:35,592] kalite: No sw_icu.js file found in locale_path` `/Users/####/.kalite/locale`
`[INFO] [2016-09-28 16:25:35,593] kalite: Writing i18nized js file to` `/Users/####/.kalite/content/locale/js/i18n/sw.js`
`[INFO] [2016-09-28 16:25:35,646] kalite: Looking for available content items.`
`[INFO] [2016-09-28 16:25:35,711] kalite: Annotating content for language: sw, channel: khan`
` `
`This may take several minutes depending on system resources...`
`[INFO] [2016-09-28 16:25:36,540] kalite: Annotation complete for language: sw, channel: khan`
`[INFO] [2016-09-28 16:25:36,541] kalite: Finished processing content pack.`

Then on restart, here's what I see (screen recorded from a Samsung Galaxy S6 with default Internet browser - same behaviour seen on Safari and Chrome on Macbook) - only the background UI is Kiswahili, but the menus are blank.


[2016_09_28_17_23_39.mp4.zip](https://github.com/learningequality/ka-lite/files/498750/2016_09_28_17_23_39.mp4.zip)

Thanks again for your time and help :)
Kevin

**Comment by mrpau-richard** [comment]
Hi @Kevsy The menu tree is empty because there's no available exercises yet, but if you download a `swahili` videos it will appear on the menu tree. We are still working on other exercises and why is not included.

![a-temp](https://cloud.githubusercontent.com/assets/8663934/18934597/2b84b54c-860d-11e6-9b38-3b0d6ba4bbe8.gif)

**Comment by Kevsy** [comment]
Thanks @mrpau-richard . Is there a location where I can bulk download the Swahili videos? Or, if not, is there a way I can download them via the browser interface, and then copy them from my local directory to a USB stick to send to the remote location (which has no Internet access)?

**Comment by mrpau-richard** [comment]
@Kevsy We have a documentation for that [here](http://ka-lite.readthedocs.io/en/latest/usermanual/userman_admin.html#bulk-video-downloads), But sadly there's no available `Swahili ` torrent file. I will try to create a torrent file for `Swahili ` and ping you when it's done.

**Comment by benjaoming** [comment]
@Kevsy - I wonder if it's because you have no videos downloaded and there aren't any translated exercises in the set.. will investigate!

**Comment by mrpau-richard** [comment]
@benjaoming based on our findings `swahili` or `sw` language code has no exercises fetch from KHAN api. We will try to use the `swa` language code for 'swahili` if it has a return exercises from KHAN api.

**Comment by benjaoming** [comment]
Hi @mrpau-richard - great find! As I commented on @mrpau-eduard 's commit, I think it might be easier to stick to our current `sw` language code and instead map our calls to upstream Khan API to `swa`.. so we don't have to change language codes in other places and so current deployments (possibly automated scripts) won't have to change as well.

**Comment by julianharty** [comment]
At the risk of derailing this topic. Can you tell me how to manage ka-lite
installed using the OSX installer please? I've installed 0.16.9 using the
OSX installer and it's running. I now need to login as an admin to install
the language pack, etc. however I can't see an easy way to do so and don't
know if there's an existing admin / super-user provided in the package.

I have found and tried the terminal interface. Here's what I get

Julians-MBP-2:~ julianharty_air$ kalite

Traceback (most recent call last):

  File "/Applications/KA-Lite/support/pyrun-2.7/share/kalite/kalitectl.py",
line 115, in <module>

    import kalite

ImportError: No module named kalite


I'm fine with raising a fresh bug for my request, however I want to get
started with trying the Swahili language pack asap.

Thanks

Julian

**Comment by mrpau-richard** [comment]
@julianharty I have the same issue with you. I just restarted my mac and it will work fine.
To use the `Swahili language pack` follow this [documentation](http://ka-lite.readthedocs.io/en/latest/faq.html?highlight=retrievecontentpack#how-can-i-install-a-language-pack-without-a-reliable-internet-connection).

@Kevsy Here's the [torrent file](https://github.com/learningequality/ka-lite/files/501071/ka-lite-0.16-resized-videos-swahili.torrent.zip) to  bulk download the Swahili videos. But it has five `Swahili` videos that's not successfully included, because it has a problem when I download it.
`

**Comment by Kevsy** [comment]
@mrpau-richard much appreciated, thanks! I'll report back once I've installed the videos.

**Comment by mrpau-richard** [comment]
@Kevsy We updated `Kiswahili` language pack with exercises and videos, you can download it [here](http://pantry.learningequality.org/downloads/ka-lite/0.16/content/contentpacks/sw.zip)

**Comment by Kevsy** [comment]
That's brilliant @mrpau-richard , thanks so much everyone for the effort! Will test this afternoon.

**Comment by Kevsy** [comment]
I'm getting a mix of English and Swahili descriptions (and sometimes category headings) - is that expected? 
![screenshot_20161007-165538](https://cloud.githubusercontent.com/assets/7903978/19197480/f56af8a8-8cb1-11e6-932c-8f0440b93998.png)
![screenshot_20161007-165654](https://cloud.githubusercontent.com/assets/7903978/19197481/f56d653e-8cb1-11e6-95a4-bf937a9faf33.png)
![screenshot_20161007-165716](https://cloud.githubusercontent.com/assets/7903978/19197482/f56f9b7e-8cb1-11e6-8f06-945fa1cfcf3d.png)

And just to check CONTENT_ROOT: I've changed this to an external drive, do the Swahili videos that you kindly provided via Torrent go into this directory, or a name subdirectory of the CONTENT_ROOT? At the moment I can't seem to pick them up from the new CONTENT_ROOT.

    kalite diagnose
    CONTENT ROOT:         /Volumes/Time Machine/kalite/content
    CONTENT SIZE:         3.7 GB

    /Volumes/Time Machine/kalite/content
    (lists a load of videos :))

**Comment by julianharty** [comment]
My understanding is that there are roughly 1200 dubbed+subtitled 'Swahili'
videos in Khan Academy. When I downloaded all the videos ka-lite has
configured I've got around 370, so far less than the 1200 (IIRC the number
comes from Erna Graz who was responsible for translating them). We
translated 100% of the ka-lite aspects, however they're dwarfed compared to
the translations needed for the khan academy sources which are also
incorporated AFAIK into ka-lite. I don't know exactly how much we managed
to translate of this body of material (it's not easy for me to determine
for a bunch of annoying reasons outside my influence) however I'd guess
somewhere between 10% and 25%. This means it's quite likely we'd get mixed
English and Swahili. There may be ways to surgically translate stuff but
that's likely to take significant effort. - Something to consider/discuss
if you and others are willing to help. I'm not able to do this alone.

**Comment by mrpau-richard** [comment]
@Kevsy Yeah, that's right because exercises, videos and topics string of the `Kiswahili` language pack is not fully translated at [Khan Academy crowdin](https://crowdin.com/project/khanacademy/sw).

**Comment by benjaoming** [comment]
@mrpau-richard - the resized torrent contains 332 mp4 files. I think what @julianharty is saying is that there should be 1,200 videos. Could it be that only dubbed videos are included in the torrent?

**Comment by julianharty** [comment]
I've found that trying to track all the videos in various languages is
time-consuming and error prone. There's a master spreadsheet in Google Docs
that 's updated from time to time (Jamie Alexandre is one of the people who
does) but it's not complete in terms of tracking all the videos. Similarly,
it's been a challenge to establish where the oracle (the official master
reference) is or whether it's maintained. During some experiments with the
Khan Academy API Joe Reeve (a friend who's been helping me from time to
time) and I found more videos but not all 1,200.

I got sufficiently frustrated last year with the mess that I eventually
parked the problem. Were some kind and knowledgable soul from Khan Academy
willing to help we might actually be able to establish a definitive
reference set of videos. For now, I can tell you that we don't have all the
videos in ka-lite, we have somewhere around 25% to 30%. Ed Resor is
currently working on trying to match up and update a copy of the master
spreadsheet. If he's successful I'll connect you so we can hopefully get
many more of the 1200 (ish) loaded.

Similarly we may need to reinvestigate what translations are downloaded
from Khan Academy and identify the more important omissions and where to
add the translations. I have some offers to help with the translations but
to be effective they need to know where to focus their attention. Someone
may then need to work with Khan Academy so we can provide the translations
(earlier this year I had problems reaching anyone there who'd help me in
this aspect).

In summary, we know we lack some translations and many of the dubbed
videos. I believe there are more available however we need to find a way to
identify and map them into the master reference so we can download
everything that has been translated. We also need an idea of where to focus
 the attention of potential translators so we can complete enough of the
translations to provide a coherent subset of the material in Swahili
(probably the more junior / simpler materials). We then need help to
facilitate the translations (potential translators for Khan Academy seem to
need to be approved by Khan Academy), get the translations done and
checked, and then finally incorporated into ka-lite, and hopefully Kolibri.

Julian

**Comment by julianharty** [comment]
I decided to have a go at finding out roughly how many videos there seem to
be in Swahili on Khan Academy. Here's what I did, the count, and some
comments based on experiments I did with Joe Reeve about 6 months ago which
indicate some of these may be in English rather than Swahili.


   1. Visit sw.khanacademy.org
   2. View page source view-source:https://sw.khanacademy.org/
   3. Search for: /video?lang=sw&format

There are 1822 matches, many more than the 322 videos in the torrent file,
or the 336 I've found in my content folder in my ka-lite 0.16.9 (installed
using pip) but perhaps a few of these are in other languages...? There are
384 according to the master Google Doc spreadsheet
https://docs.google.com/spreadsheets/d/1haV0KK8313lG-_Ay2REplQuMquRStZumB3zxmmtYqO0/edit?usp=sharing

However.... When Joe Reeve and I spent a day or so earlier this year trying
to fill gaps in the master Google Doc spreadsheet of ka-lite videos in
various languages we discovered some of the 'missing' videos available from
Khan Academy's API (IIRC) were actually in English. I can't remember how
many but we encounter them fairly often during our sampling.

All this leads me to guess that there should be significantly more than 332
videos available - at least 384 unless there are known problems with some
of the ones listed in the current spreadsheet. There may be 1800+ videos in
Swahili (there maybe some on YouTube not listed on the sw.khanacademy.org
page) or - more likely somewhere between these 2 amounts - I tend to revert
back to the working figure of around 1200 while hoping there'll be
significantly more available, nonetheless YouTube claims there are 913
videos (see attached screenshot taken from YouTube's search results).

More info: the main YouTube channel seems to be
https://www.youtube.com/channel/UCZAYCR3NauQk8jP--Iok-Fw According to
YouTube the most recent video was uploaded 3 years ago! They have low view
counts, the most viewed video has less than 1200 views.

So, in summary there doesn't seem to be a definitive answer, however we
should be able to find and include more than 332 videos in Swahili, I hope
:)

**Comment by julianharty** [comment]
BTW: is the online editing site for translations still available - what
used be http://kitchen.learningequality.org:1187 For those of you who've
not used or seen this site, it allowed people to select a language (Swahili
is the obvious choice for this discussion) and view the UI of ka-lite in a
way that people could then click on text (and particularly
untranslated content) and be taken to the correct location on the
crowdin.com online translation site used by ka-lite and khan academy.
(Translations for khan academy seem to also use other tools these days,
I've lost track a bit, but Jamie demoed a tool I hadn't seen or used). This
'kitchen' web site might help us find and fix at least some of the more
important translations in the menus, etc.

**Comment by mrpau-richard** [comment]
@julianharty & @benjaoming  The videos inside in `Kiswahili` torrent file is mixed, from khan api and doc spreadsheet and other videos is not included in torrent file because it has a problem when I manually downloading it in KA Lite.

When I filtered the `Kiswahili` videos in the content pack database I got 342 videos in result, but videos loaded in KA Lite is only 334 It's because some youtube_id has value "#NAME?".

![screen shot 2016-10-10 at 11 50 35 am](https://cloud.githubusercontent.com/assets/8663934/19226604/d27b4fa2-8edf-11e6-81f1-ed5413df9c7e.png)

![screen shot 2016-10-10 at 11 49 49 am](https://cloud.githubusercontent.com/assets/8663934/19226597/bab094ea-8edf-11e6-9c85-b7394e9660da.png)

![screen shot 2016-10-10 at 11 29 21 am](https://cloud.githubusercontent.com/assets/8663934/19226612/dec92716-8edf-11e6-992e-a9d9979e2d4c.png)


I also think why we have  only 342 videos in the swahili content pack database is some "Kiswahili" youtube_ids in doc spreadsheet has no video entry in English. Because we are using English videos to find "Kiswahili" videos equivalent in doc spreadsheet.

**Comment by cpauya** [comment]
Hi @julianharty - thanks much for validating the Swahili videos on KA.  It's very helpful.

To expound on what @mrpau-richard said:  There are `youtube_ids` from the KA API calls that were included in the returned JSON but are still tagged as English via the `translated_video_lang` key.  Since we can't include that item in the list, what we did was lookup these `youtube_ids` to the **Dubbed Video Mappings** at https://docs.google.com/spreadsheets/d/1k5xh2UXV3EchRHnYzeP6-YGKrmba22vsltGSuU9bL88/ under the Kiswahili column and when found, use that and then change the `translated_video_lang` into `sw`.  @mrpau-richard please feel free to correct me here if I missed something. :)

We're now checking how many look-ups do we have to the Dubbed Video Mappings so we can determine if it coincides with the number of videos you reported based on Youtube playlist or KA.  Those Youtube videos sure are outdated (3 years?), but hope they are still applicable.

Initially, our recommendation is to manually update the **Dubbed Video Mappings** spreadsheet by looking up from the Youtube playlist you mentioned so that our list of Swahili videos in the content pack will increase.

**Comment by jamalex** [comment]
@cpauya What do you do in the case that a YT ID isn't found either in the KA API or in the spreadsheet? Presumably the desired behaviour would then be to skip it -- perhaps now it's just falling back to the EN video?

**Comment by jamalex** [comment]
Building on what @julianharty showed, I ran some further stats.

Looking at https://sw.khanacademy.org/, we do see the quoted number if we look at all video links:
`len(video_ids) => 1823`

However, if we uniquify, we get a much smaller number, though still greater than what we're seeing elsewhere:
`len(set(video_ids)) => 740`
(in part, this is because every link appears at least twice -- probably for different screen layouts -- and then some appear 4 times because they're in multiple playlists or whatnot)

One reason we're probably getting less than 384 (as shown in the spreadsheet), in addition to the #NAME stuff, is that KA may have removed/replaced some of the original EN YouTube videos, so they're no longer in the topic tree. Since we're using the EN topic tree and mapping videos into it, those ones won't be included.

It looks like some manual work will be needed, in terms of mapping, to make progress on including more Swahili videos. Probably easier doing this in the spreadsheet for now, though doing it in the KA DB so it feeds into the API would be more sustainable.

**Comment by mrpau-richard** [comment]
Hi @jamalex The behavior in case the `youtube_ids` isn't found in either in the KA API or in the spreadsheet is it will not return any videos. And it will not appear in the KA Lite topic tree.

I also agree, to manually add the `youtube_ids` that's not added in the spreadsheet.

**Comment by jamalex** [comment]
@mrpau-richard Thanks for the clarification. Going back to the first message in this thread, it's not clear to me why there would be English videos getting included, then. We only use the results from the API if the EN id doesn't match the translated ID, and the fallback IDs in the Swahili column should all be in Swahili.

@Kevsy @julianharty Can you list a few examples of videos that are still in English, so the team can look into where those IDs are coming from? Thanks!

**Comment by Kevsy** [comment]
@jamalex yes will do - we're shipping the USB of videos from the torrent to Tanzania, and will report back from the local team. I've also asked for an identification of the core 'primary school' learning path so that we can identify the priority translations; and ideally we can help with translating any gaps via crowdin. Thanks everyone for all the help!!

**Comment by mrpau-eduard** [comment]
Hi @jamalex I file an issue #5328  for the videos that are in the French packs are in English.
I list the youtube ids and the urls where the English videos  is playing instead of French videos.

**Comment by benjaoming** [comment]
A new mechanism for retrieving titles and descriptions via Youtube API has landed.

@mrpau-eduard could you say how this affects the Swahili content pack and if a new one will be uploaded in this regard? Thanks :)

**Comment by benjaoming** [comment]
@Kevsy @julianharty - please feel invited to try out the prerelease 0.17.0b5 described in #5390 

The new Swahili content packs are ready, and they should also provide you with the right videos.

**Comment by jamalex** [comment]
Note that there's a torrent file for 0.17 that also includes all the vids for 0.16: http://pantry.learningequality.org/downloads/ka-lite/0.17/content/ka-lite-0.17-resized-videos-swahili.torrent

**Comment by benjaoming** [comment]
I'm going to close this issue, assuming that a lot of ground has been covered in retrieving the right data from the Youtube API, moreover that information has been sparse, @Kevsy and @julianharty haven't reported back since 2016.

We are rebuilding content packs again this next week.

A new triage of Swahili content is welcomed, but best to put that in a new Issue.

Thanks all for your work on this!!
